external-crates/move/move-compiler/src/cfgir/visitor.rs

```diff
@@ -0,0 +1,460 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::BTreeMap, fmt::Debug};
+
+use crate::{
+    cfgir::{
+        absint::{AbstractDomain, AbstractInterpreter, JoinResult, TransferFunctions},
+        cfg::BlockCFG,
+        CFGContext,
+    },
+    command_line::compiler::Visitor,
+    diagnostics::{Diagnostic, Diagnostics},
+    hlir::ast::{
```

Comment:
Since this visitor operates on `hlir` anyways, I was wondering if it would be possible (and would makse sense) to expose the entire AST for a module to it. It could make some things simpler - for example we would have access to all types declared in a given module.

Reply:
It would be tricky, but doable. 
The compiler destroys/moves the AST when translating from one step to another. So parts of the hlir ast are gone by the time this happens. 

I can try reworking it in a separate PR though. It would just require some restructuring

Reply:
It required a lot of fun refactoring, but I got it to work!